### ISSUE TITLE
[BUG] - Auto tunnelling does not switch on the tunnel

### ISSUE THREAD

**Issue by etofi** [zaneschepke/wgtunnel#127]
**Describe the bug**
When I leave my house and the trusted WiFi, Auto Tunnelling does not switch the tunnel on. Only when I open the app.
I have set everything correctly, allowed permanent location access and switched off battery optimisation.

**Smartphone (please complete the following information):**

- Device: Motorola Edge 20
- Android Version: Android 13
- App Version 3.3.5

**To Reproduce**

1. Leave trusted WIFI

**Expected behavior**
Auto tunnelling should switch on the tunnel when you leave the trusted WiFi even without having to open the app. 

**Screenshots (Only if necessary)**

**Additional context**

**Comment by zaneschepke** [comment]
Do you have tunnel on mobile data enabled?

**Comment by zaneschepke** [comment]
Thank you for the screenshot! A few more follow up questions..

Does this happen every time you close the app or just sometimes? 

Is the auto tunnel service in paused state on the main screen?

Could you try starting auto tunneling, closing the app, turning off your wifi on your phone, and then see if it auto tunnels?

Basically, I am trying to determine if this failure only happens after the auto tunnel service has been running for a while (maybe the system killed it) or if it happens consistently every time you test.

**Comment by etofi** [comment]
> Does this happen every time you close the app or just sometimes?

Just sometimes. 

> Is the auto tunnel service in paused state on the main screen?

State is active 

> Could you try starting auto tunneling, closing the app, turning off your wifi on your phone, and then see if it auto tunnels?

In the moment it works. But sometimes not. I can not reproduce it. 

> Basically, I am trying to determine if this failure only happens after the auto tunnel service has been running for a while (maybe the system killed it) or if it happens consistently every time you test.

I think it is a "run for a while" problem

**Comment by etofi** [comment]
![Screenshot_20240303-163111](https://github.com/zaneschepke/wgtunnel/assets/19774831/d405f765-5ed7-46eb-b9a6-5a0c731eeb92)
So today is the day again. The tunnel is not switched on even though I've been travelling for over 3 hours now...
Status is active.
The tunnel is activated immediately when the app is opened

**Comment by slubman** [comment]
I have the same issue using the Play store version, with a Pixel 8.

The auto-tunnel will not connect the tunnel when I disconnecte from wifi or when I leave wifi coverage.
If I open the application the connection is established quickly. And the auto-connection also works for some time after that.

Is there a way to enable some logging to understand what is going on ?

**Comment by zaneschepke** [comment]
> I have the same issue using the Play store version, with a Pixel 8.
> 
> The auto-tunnel will not connect the tunnel when I disconnecte from wifi or when I leave wifi coverage. If I open the application the connection is established quickly. And the auto-connection also works for some time after that.
> 
> Is there a way to enable some logging to understand what is going on ?

Hello! Thank you for the additional info. I'm sorry for the delay. I should have a fix out for this soon. Additionally, the next version of the app I am adding a logs screen to view the logs.

**Comment by etofi** [comment]
I have seen that 3.3.9 is now available. 
When will it be available in F-Droid?

**Comment by zaneschepke** [comment]
> I have seen that 3.3.9 is now available. When will it be available in F-Droid?

Hey! It usually takes about a week, but there are currently some issues with Fdroid pipeline because I have a lib that is still in process of getting published to maven central. It is currently hosted on my own package repository, but fdroid doesn't support this. Current issues for this is #47

**Comment by licaon-kter** [comment]
F-Droid is currently upgrading to Debian Bookworm which has Java 17 installed by default.

It will take a cycle or two, so breakage is expected until done.

**Comment by kiekerjan** [comment]
I'm seeing the same issue. The tunnel is not always automatically enabled on entering untrusted wifi or mobile data. For me too it seems like a "run for a while" problem. When I start the WG tunnel app, it immediately enables the tunnel. 
I have set battery usage to "unlimited" (as opposed to "optimised" or "reduced")
I'm on the 3.4.2 version from the play store.
I don't think the logs will be helpful, as when I look at "read the logs" they only go back to after I started the app, so don't show the time period in which autotunnel did not trigger.

**Comment by zaneschepke** [comment]
> I'm seeing the same issue. The tunnel is not always automatically enabled on entering untrusted wifi or mobile data. For me too it seems like a "run for a while" problem. When I start the WG tunnel app, it immediately enables the tunnel. I have set battery usage to "unlimited" (as opposed to "optimised" or "reduced") I'm on the 3.4.2 version from the play store. I don't think the logs will be helpful, as when I look at "read the logs" they only go back to after I started the app, so don't show the time period in which autotunnel did not trigger.

I'll keep investigating. This is a very hard one to troubleshoot. 

Do you know if the auto tunneling notification was still visible when you noticed this issue or was it closed? 

Also, do you have any rough estimate on how often this occurs so I can try to replicate on a few different devices?

**Comment by kiekerjan** [comment]
I probably swiped that icon away manually. I'll keep it present and see what happens.
Rough estimate is like 1 in 3 times I leave my home (where the trusted wifi is)

**Comment by kiekerjan** [comment]
So it just happened once more. The icon reporting "monitoring network state changes: active" was still present. When I clicked it, the WG tunnel UI was opened and it resumed auto tunneling.
The logging went only back to the moment when I opened the UI.

**Comment by slubman** [comment]
> So it just happened once more. The icon reporting "monitoring network state changes: active" was still present. When I clicked it, the WG tunnel UI was opened and it resumed auto tunneling. The logging went only back to the moment when I opened the UI.

I got exactely the same, auto-tunnel did not activate, when opening the application immediatly re-established the connection, the log only start from when I opened the application, despite the notification being shown.

**Comment by genomez** [comment]
> > So it just happened once more. The icon reporting "monitoring network state changes: active" was still present. When I clicked it, the WG tunnel UI was opened and it resumed auto tunneling. The logging went only back to the moment when I opened the UI.
> 
> I got exactely the same, auto-tunnel did not activate, when opening the application immediatly re-established the connection, the log only start from when I opened the application, despite the notification being shown.

I'm seeing the same thing. It happened to me once yesterday and once today. The toggle indicated I was connected, but I clearly wasn't. Once I opened the app, I was immediately connected. Both times the auto tunneling notifications were visible. 

Thanks for the great app!

**Comment by dkoppenh** [comment]
I have the same issue. After a time, even though the notifications are active, the tunnel doesn't start when it should.
Once, I just had to open the app for the tunnel to activate, but more recently, I had to manually stop and start the auto-tunnelling.

A point of reference is that I changed the notifications to "silent" in the Android notification settings.

**Comment by zaneschepke** [comment]
> I have the same issue. After a time, even though the notifications are active, the tunnel doesn't start when it should. Once, I just had to open the app for the tunnel to activate, but more recently, I had to manually stop and start the auto-tunnelling.
> 
> A point of reference is that I changed the notifications to "silent" in the Android notification settings.

I'm still working on getting to the bottom of this one. This is a top priority.

**Comment by zaneschepke** [comment]
I made some changes to the latest release 3.4.3 that could potentially help with this issue. Please let me know if anyone notices any differences or if this issue is still persisting.

**Comment by etofi** [comment]
Unfortunately, only version 3.4.2 is available on F-Droid. The app has now been updated to 3.4.4...

**Comment by zaneschepke** [comment]
> Unfortunately, only version 3.4.2 is available on F-Droid. The app has now been updated to 3.4.4...

Yeah, Fdroid build/deployment pipelines takes up to one week. If you want updates immediately, feel free to add my custom Fdroid repo to the Fdroid app. 

https://github.com/zaneschepke/fdroid

**Comment by etofi** [comment]
> Yeah, Fdroid build/deployment pipelines takes up to one week. If you want updates immediately, feel free to add my custom Fdroid repo to the Fdroid app.
> 
> https://github.com/zaneschepke/fdroid

In your custom Repo only version 3.4.2 is available, too via F-Droid app.

**Comment by zaneschepke** [comment]
> > Yeah, Fdroid build/deployment pipelines takes up to one week. If you want updates immediately, feel free to add my custom Fdroid repo to the Fdroid app.
> > 
> > https://github.com/zaneschepke/fdroid
> 
> In your custom Repo only version 3.4.2 is available, too via F-Droid app. 

Hmm. Are you sure you have the repo added properly? I'm able to install 3.4.4 with my repo via F-droid.

**Comment by slubman** [comment]
> Hmm. Are you sure you have the repo added properly? I'm able to install 3.4.4 with my repo via F-droid. 

I had to uninstall the F-Droid version to be able to install the one from the WGTunnel repository.

**Comment by kladderadeng** [comment]
I have the same issue that the Auto connect / disconnect is not doing anything for me. I installed the Store version on LineageOS, enabled "Tunnel on mobile data", enabled "Tunnel on untrusted wifi" and added my WiFi SSID. I also tried with setting a default tunnel and not (I only have one tunnel anyway).
I allowed the app to be exempt from the battery savings and allowed it access to my location data. I restarted the Smartphone, tried with the app opened, in background and closed.

My tests was always done by turning the wifi off and on and waiting a minute. Maybe I am testing wrong and wifi should be enabled, but you have to leave the wifi zone?

**Comment by zaneschepke** [comment]
> I have the same issue that the Auto connect / disconnect is not doing anything for me. I installed the Store version on LineageOS, enabled "Tunnel on mobile data", enabled "Tunnel on untrusted wifi" and added my WiFi SSID. I also tried with setting a default tunnel and not (I only have one tunnel anyway). I allowed the app to be exempt from the battery savings and allowed it access to my location data. I restarted the Smartphone, tried with the app opened, in background and closed.
> 
> My tests was always done by turning the wifi off and on and waiting a minute. Maybe I am testing wrong and wifi should be enabled, but you have to leave the wifi zone?

This specific use case has a bug that I'm currently working on where it does not turn on the tunnel when switching from wifi to mobile data. I should have a fix out for this soon!

**Comment by KjartanOli** [comment]
Not sure if this belongs here or in a new issue but I'm having similar problems with auto-tunneling.   In my case I want to connect to a specific tunnel when I connect to a specific WiFi network.  I've added the SSID to the 'Use tunnel on wifi name' field on the settings page for the tunnel and enabled auto-tunneling but when I connect to that WiFi network nothing happens.

**Comment by zaneschepke** [comment]
> Not sure if this belongs here or in a new issue but I'm having similar problems with auto-tunneling. In my case I want to connect to a specific tunnel when I connect to a specific WiFi network. I've added the SSID to the 'Use tunnel on wifi name' field on the settings page for the tunnel and enabled auto-tunneling but when I connect to that WiFi network nothing happens.

I've just released a new version 3.4.5. Could you try this on that version and see if you are getting the same behavior? Also, could you please share your phone model and version? 

Please make sure you also have "tunnel on untrusted wifi" enabled on the settings screen and auto tunneling started and not in the paused state.

**Comment by KjartanOli** [comment]
Turning on 'tunnel on untrusted wifi' makes it work.  The tunnel activates when connect to the WiFi network.  However, I don't want to tunnel on all untrusted networks, currently this is the only network I want to tunnel on.  Is that possible?

**Comment by zaneschepke** [comment]
> Turning on 'tunnel on untrusted wifi' makes it work.  The tunnel activates when connect to the WiFi network.  However, I don't want to tunnel on all untrusted networks, currently this is the only network I want to tunnel on.  Is that possible?

Not currently possible, but I have had a request for a similar feature so I'll work on adding this.

**Comment by KjartanOli** [comment]
Thank you.

**Comment by tuxArg** [comment]
Hi, I'm seeing this same issue. Running 3.4.5
I have 2 tunnels configured, both are the same except that of them has an extra peer.
I want to be able to autoconnect to tunnel 1 if I'm connected to SSID A. And use tunnel 2 in any other case (another wifi or mobile data).
So I added SSID A to tunnel 1.
And mobile data and default tunnel checkbox to tunnel 2.

In general tab,
I checked "untrusted wifis" (leaving blank the trusted ones), mobile data and ethernet.

When I start auto-tunnel it connects to the tunnel it should. But if network changes, it just disconnects until I open the app again and only then it autoselects the right tunnel.

And other issues:
There are 2 active notifications, are both of them neccesary? Any how I could hide them?

In my use case that I want a persistent vpn, why can't I check always on vpn?

Thanks

**Comment by zaneschepke** [comment]
> Hi, I'm seeing this same issue. Running 3.4.5 I have 2 tunnels configured, both are the same except that one of them has an extra peer. I want to be able to autoconnect to tunnel 1 if I'm connected to SSID A. And use tunnel 2 in any other case (another wifi or mobile data). So I added SSID A to tunnel 1. And mobile data and default tunnel checkbox to tunnel 2.
> 
> In general tab, I checked "untrusted wifis" (leaving blank the trusted ones), mobile data and ethernet.
> 
> When I start auto-tunnel it connects to the tunnel it should. But if network changes, it just disconnects until I open the app again and only then it autoselects the right tunnel.
> 
> And other issues: There are 2 active notifications, are both of them neccesary? Any how I could hide them?
> 
> In my use case that I want a persistent vpn, why can't I check always on vpn?
> 
> Thanks

I am able to reproduce this issue. I working on releasing a fix right now. 

As for the other issue: One notification is to persist the vpn service the other is for the service watching network/state changes for auto tunneling. On Android 14, these persistent notifications can be dismissed.

You can still enable AOVPN in Android setting still, that setting just doesn't allow AOVPN to control bringing up the tunnel so it does not conflict with auto tunneling.

**Comment by zaneschepke** [comment]
> Hi, I'm seeing this same issue. Running 3.4.5 I have 2 tunnels configured, both are the same except that one of them has an extra peer. I want to be able to autoconnect to tunnel 1 if I'm connected to SSID A. And use tunnel 2 in any other case (another wifi or mobile data). So I added SSID A to tunnel 1. And mobile data and default tunnel checkbox to tunnel 2.

Please give this version and shot and see if it fixes the issues you are seeing.
 
https://github.com/zaneschepke/wgtunnel/releases/tag/3.4.6

**Comment by tuxArg** [comment]
@zaneschepke 3.4.6 is working for me!!

Regarding the notifications, using the _official_ wireguard client I am able to hide it if setting AOVPN. Why is it different here?

**Comment by zaneschepke** [comment]
I'll take a look. There would still need to be at least one for the auto tunnel service.

**Comment by kiekerjan** [comment]
Just came by to mention that in the last few weeks I didn’t see the issue occuring. Difficult to prove absence of something, but it's looking good from where I'm sitting.

**Comment by kafisc1** [comment]
> I'm still having the issue. Sitting at the dentist office right now on 5G and the VPN has not connected but is monitoring. Just opened the app right now though and it immediately switched on.

I can confirm this behavior when I'm at work. I will try to create some logs tomorrow.

**Comment by dkoppenh** [comment]
Strangely enough, I had the opposite problem today. I connected to a non trusted wifi, and the tunnel came up. 

When I reconnected to my trusted wifi, the tunnel stayed up. As soon as I opened the app, it correctly disconnected the tunnel.

**Comment by kafisc1** [comment]
Today I was able to reproduce the issue and created logs. 
I left home WiFi at 05:52 o'clock an wgtunnel connected correctly trough mobile network.
I arrived at work around 06:03 o'clock, entered the WiFi at work automatically and wgtunnel was not able to connect.
Please see attached logs for further information.
Thank you in advance
[wg_tunnel_logs-1720498108.txt](https://github.com/user-attachments/files/16138030/wg_tunnel_logs-1720498108.txt)

**Comment by zaneschepke** [comment]
Please try this prerelease build that I'm hoping fixes this issye: https://github.com/zaneschepke/wgtunnel/releases/tag/auto-tunnel-reliability-testing

**Comment by dkoppenh** [comment]
I hate to say this, but I'm running 3.5.1 and the auto tunneling failed this evening.  I opened the app, and only then did the tunnel came up.

On the bright side, the problem seems to happen less on recent versions.

**Comment by dkoppenh** [comment]
Today, it auto connected just fine. I'll try to pay more attention to the state of my phone and app before I expect to have to auto connect next time.

**Comment by dkoppenh** [comment]
I just turned off my wifi, after seeing the notification that auto tunneling was active via the notification. It should have connected to VPN over my cell data connection.

The VPN didn't activate (twice, I even toggled the wifi again) until I opened the app.

I tried to press the save icon in the app to save my logs, but then the app closed (notification still there), then I tried saving logs again and the whole app & notification crashed.

I did recently switch to using Amnezia parameters for connecting to my non-AmneziaWG server.
